File: problems/756_pyramid_transition_matrix/Solution.java
import java.util.*;

class Solution {
    Map<String, List<Character>> mapping = new HashMap<>();

    public boolean pyramidTransition(String bottom, List<String> allowed) {
        for (String s : allowed)
            mapping.computeIfAbsent(s.substring(0,2), k -> new ArrayList<>()).add(s.charAt(2));
        return solve(bottom, new StringBuilder(), 0);
    }

    private boolean solve(String bottom, StringBuilder next, int idx) {
        if (bottom.length() == 1) return true;
        if (idx == bottom.length() - 1) return solve(next.toString(), new StringBuilder(), 0);
        String key = bottom.substring(idx, idx + 2);
        for (char c : mapping.getOrDefault(key, new ArrayList<>())) {
            next.append(c);
            if (solve(bottom, next, idx + 1)) return true;
            next.deleteCharAt(next.length() - 1);
        }
        return false;
    }
}

File: problems/756_pyramid_transition_matrix/solution.py
from typing import List


class Solution:
    def pyramidTransition(self, bottom: str, allowed: List[str]) -> bool:
        mapping = collections.defaultdict(list)
        for s in allowed:
            mapping[(s[0], s[1])].append(s[2])
        def solve(bottom, next_row, idx):
            if len(bottom) == 1:
                return True
            if idx == len(bottom) - 1:
                return solve(next_row, '', 0)
            for c in mapping.get((bottom[idx], bottom[idx+1]), []):
                if solve(bottom, next_row + c, idx + 1):
                    return True
            return False
        return solve(bottom, '', 0)
